src/kit_version.rs

```diff
@@ -0,0 +1,207 @@
+                };
+                let major = match major {
+                    Ok(e) => e,
+                    Err(_) => return Err("Unable to parse version major value"),
+                };
+                let minor = match minor {
+                    Ok(e) => e,
+                    Err(_) => return Err("Unable to parse version minor value"),
+                };
+                let patch = match patch {
+                    Ok(e) => e,
+                    Err(_) => return Err("Unable to parse version patch value"),
+                };
+
+                // Check the dev indicator
+                let dev = captures.get(5).is_some();
+
+                let commit = captures.get(6);
+                let branch = captures.get(7);
+
+                let build_info = match (commit, branch) {
+                    (None, None) => None,
+                    (None, Some(_)) => None,
+                    (Some(commit), None) => Some(BuildInfo {
+                        commit: commit.as_str().to_string(),
+                        branch: None,
+                    }),
+                    (Some(commit), Some(branch)) => Some(BuildInfo {
+                        commit: commit.as_str().to_string(),
+                        branch: Some(branch.as_str().to_string()),
+                    }),
+                };
+
+                Ok(KitVersion {
+                    epoch,
+                    major,
+                    minor,
+                    patch,
+                    dev,
+                    build_info,
+                })
+            }
+            None => Err("version was not in valid format."),
+        }
+    }
+}
+
+impl Serialize for KitVersion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let s = format!("{}", self);
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for KitVersion {
+    fn deserialize<D>(deserializer: D) -> Result<KitVersion, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(KitVersionVisitor)
+    }
+}
+
+struct KitVersionVisitor;
+
+impl<'de> Visitor<'de> for KitVersionVisitor {
+    type Value = KitVersion;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a valid kit version string")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        match KitVersion::try_from(value) {
+            Ok(s) => Ok(s),
+            Err(e) => Err(de::Error::custom(e)),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::KitVersion;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn parse_version_normal() {
+        KitVersion::try_from("2021.0.0.1").unwrap();
+    }
+
+    #[test]
+    fn parse_version_dev() {
+        KitVersion::try_from("2021.0.0.1dev").unwrap();
```

Comment:
Nit: Using the `assert` macro is more idiomatic here than unwrapping. `.is_some` can be useful for that.